packages/@uppy/url/src/Url.tsx

```diff
@@ -0,0 +1,238 @@
+        fill="#FF753E"
+        fill-rule="nonzero"
+      />
+    </svg>
+  )
+}
+
+function addProtocolToURL(url: string) {
+  const protocolRegex = /^[a-z0-9]+:\/\//
+  const defaultProtocol = 'http://'
+  if (protocolRegex.test(url)) {
+    return url
+  }
+
+  return defaultProtocol + url
+}
+
+function canHandleRootDrop(e: DragEvent) {
+  const items = toArray(e.dataTransfer!.items)
+  const urls = items.filter(
+    (item) => item.kind === 'string' && item.type === 'text/uri-list',
+  )
+  return urls.length > 0
+}
+
+function checkIfCorrectURL(url?: string) {
+  if (!url) return false
+
+  const protocol = url.match(/^([a-z0-9]+):\/\//)![1]
+  if (protocol !== 'http' && protocol !== 'https') {
+    return false
+  }
+
+  return true
+}
+
+function getFileNameFromUrl(url: string) {
+  const { pathname } = new URL(url)
+  return pathname.substring(pathname.lastIndexOf('/') + 1)
+}
+
+/*
+ * Response from the /url/meta Companion endpoint.
+ * Has to be kept in sync with `getURLMeta` in `companion/src/server/helpers/request.js`.
+ */
+type MetaResponse = {
+  name: string
+  type: string
+  size: number | null
+  statusCode: number
+}
+
+export type UrlOptions = CompanionPluginOptions
+
+export default class Url<M extends Meta, B extends Body> extends UIPlugin<
+  UrlOptions,
+  M,
+  B
+> {
+  static VERSION = packageJson.version
+
+  static requestClientId = Url.name
+
+  icon: () => JSX.Element
+
+  hostname: string
+
+  client: RequestClient<M, B>
+
+  canHandleRootDrop: typeof canHandleRootDrop
+
+  constructor(uppy: Uppy<M, B>, opts: UrlOptions) {
+    super(uppy, opts)
+    this.id = this.opts.id || 'Url'
+    this.title = this.opts.title || 'Link'
+    this.type = 'acquirer'
+    this.icon = () => <UrlIcon />
+
+    // Set default options and locale
+    this.defaultLocale = locale
+
+    this.i18nInit()
+
+    this.hostname = this.opts.companionUrl
+
+    if (!this.hostname) {
+      throw new Error(
+        'Companion hostname is required, please consult https://uppy.io/docs/companion',
+      )
+    }
+
+    // Bind all event handlers for referencability
+    this.getMeta = this.getMeta.bind(this)
+    this.addFile = this.addFile.bind(this)
+    this.handleRootDrop = this.handleRootDrop.bind(this)
+    this.handleRootPaste = this.handleRootPaste.bind(this)
+
+    this.client = new RequestClient(uppy, {
+      pluginId: this.id,
+      provider: 'url',
+      companionUrl: this.opts.companionUrl,
+      companionHeaders: this.opts.companionHeaders,
+      companionCookiesRule: this.opts.companionCookiesRule,
+    })
+
+    this.uppy.registerRequestClient(Url.requestClientId, this.client)
+  }
+
+  async getMeta(url: string): Promise<MetaResponse> {
+    try {
+      return this.client.post<MetaResponse>('url/meta', { url })
+    } catch (error) {
+      this.uppy.log('[URL] Error:')
+      this.uppy.log(error)
+      throw new Error('Failed to fetch the file')
+    }
```

Comment:
This changed a lot from the previous implementation, are sure it wasn't make like this on purpose, to support a more ancient version of Companion that did return code 200 with a `error` key? I think it should be reverted, with maybe a TODO to remove the handler (the rethrow of an error is not useful).
```suggestion
  private getMeta = (url: string): Promise<MetaResponse> => {
    return this.client.post<MetaResponse>('url/meta', { url }).then((res) => {
      // TODO: remove this handler in the next major
      if ((res as any).errors) {
        this.uppy.log('[URL] Error:')
        this.uppy.log(error)
        throw new Error('Failed to fetch the file')
      }
      return res
    })
```